src/storage/mvcc/reader.rs

```diff
@@ -401,9 +406,158 @@ impl MvccReader {
                     },
                 }
             };
-            if let Some(v) = self.get(&key, ts)? {
+
+            if let Some(v) = self.reverse_get_impl(&key, ts)? {
                 return Ok(Some((key, v)));
             }
+
+            // reverse_get_impl may call write_cursor's prev.
+            if !self.write_cursor.as_mut().unwrap().valid() {
+                write_valid = false;
+            }
+        }
+    }
+
+    fn reverse_get_impl(&mut self, user_key: &Key, ts: u64) -> Result<Option<Value>> {
+        assert!(self.lock_cursor.is_some());
+        assert!(self.write_cursor.is_some());
+
+        // Check lock.
+        match self.isolation_level {
+            IsolationLevel::SI => {
+                let lock = {
+                    let l_cur = self.lock_cursor.as_mut().unwrap();
+                    if l_cur.valid() && l_cur.key() == user_key.encoded().as_slice() {
+                        self.statistics.lock.processed += 1;
+                        Some(Lock::parse(l_cur.value())?)
+                    } else {
+                        None
+                    }
+                };
+                if let Some(l) = lock {
+                    self.check_lock_impl(user_key, ts, l)?;
+                }
+            }
+            IsolationLevel::RC => {}
+        }
+
+        // Get value for this user key.
+        let mut round = 0;
+        let mut lastest_version = (None /* start_ts */, None /* short value */);
```

Comment:
Maybe `latest_statts`,`latest_short_value` will be better to understand.

Reply:
They are combined because if you change one of them, the other one must be changed. If we use two variables, it will bring some potential risk that only change one of them.